### PR TITLE
pthread: fix pthread_once mutex locking

### PR DESCRIPTION
**Pull request by lukileczo**
JIRA: RTOS-288

## Description

fixes `pthread_once` deadlocking
## Motivation and Context

JIRA: RTOS-288
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: `ia32-generic-qemu`

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: include/sys/types.h [hunk 1]
```diff
@@ -85,7 +85,11 @@ typedef struct pthread_condattr_t {
 
 typedef struct __pthread_key_t *pthread_key_t;
 
-typedef uint32_t pthread_once_t;
+typedef struct pthread_once_t {
+	uint32_t once;
+	handle_t lock;
+	volatile int initialized;
```
Comment (agkaminski) [inline include/sys/types.h:91]: Maybe use pthread mutex here and use lazy init code already in place for it?

**Review by agkaminski — COMMENTED** [review]
Is there any way to free resource (lock) after this? I see that it's not freed right now

File: include/sys/types.h [hunk 2]
```diff
@@ -85,7 +85,10 @@ typedef struct pthread_condattr_t {
 
 typedef struct __pthread_key_t *pthread_key_t;
 
-typedef uint32_t pthread_once_t;
+typedef struct pthread_once_t {
```
Comment (nalajcie) [inline include/sys/types.h:88]: should be unnamed struct or name it differently than the type

**Review by agkaminski — APPROVED** [review]

**Comment by lukileczo** [comment]
> can creating new mutex every time when `pthread_once` is called be avoided (especially the most common path - the `init_routine` was already called and finished)?

I don't think it's possible, if we were to check `once_control->once` we'd have to do it under mutex specific to this `pthread_once` (which may not exist at this point) or not destroying the mutex on exit, which results in leaking this resource

**Comment by nalajcie** [comment]
> > can creating new mutex every time when `pthread_once` is called be avoided (especially the most common path - the `init_routine` was already called and finished)?
> 
> I don't think it's possible, if we were to check `once_control->once` we'd have to do it under mutex specific to this `pthread_once` (which may not exist at this point) or not destroying the mutex on exit, which results in leaking this resource

You sure about that? :)

https://github.com/lattera/glibc/blob/master/nptl/pthread_once.c#L139

IMHO using atomic variables in pthreads generally would make the implementation faster but that's a separate topic.

If we don't want to use atomic variables right now we should be able to do it with global lock and three-state once variable (similarly to `glibc` implementation)

**Review by agkaminski — APPROVED** [review]

**Pull request merged**